scrapy/core/downloader/handlers/negotiation.py

```diff
@@ -0,0 +1,259 @@
+from scrapy.core.http2.protocol import H2ClientProtocol
+from scrapy.http.request import Request
+from scrapy.settings import Settings
+from scrapy.spiders import Spider
+from scrapy.utils.datatypes import LocalCache
+from scrapy.utils.python import to_bytes
+
+SUPPORTED_PROTOCOLS = ('h2', 'http/1.1')
+
+
+@implementer(IHandshakeListener)
+class NegotiateProtocol(Protocol):
+    def __init__(self, negotiated_protocol_deferred: Deferred):
+        self._negotiated_protocol_deferred = negotiated_protocol_deferred
+
+    def connectionLost(self, reason=connectionDone):
+        if not self._negotiated_protocol_deferred.called:
+            self._negotiated_protocol_deferred.errback(reason)
+
+    def handshakeCompleted(self) -> None:
+        self._negotiated_protocol_deferred.callback(self.transport)
+
+
+class NegotiateProtocolFactory(Factory):
+    def __init__(
+        self,
+        acceptable_protocols: List[bytes],
+        negotiated_protocol_deferred: Deferred
+    ) -> None:
+        self.acceptable_protocols = acceptable_protocols
+        self.negotiated_protocol_deferred = negotiated_protocol_deferred
+
+    def buildProtocol(self, addr) -> NegotiateProtocol:
+        return NegotiateProtocol(self.negotiated_protocol_deferred)
+
+    def acceptableProtocols(self) -> List[bytes]:
+        return self.acceptable_protocols
+
+
+class NegotiateAgent:
+    def __init__(
+        self, reactor: ReactorBase,
+        acceptable_protocols=SUPPORTED_PROTOCOLS,
+        context_factory=BrowserLikePolicyForHTTPS(),
+        connect_timeout: Optional[float] = None, bind_address: Optional[bytes] = None
+    ) -> None:
+        self._reactor = reactor
+        self._acceptable_protocols = acceptable_protocols
+        self._context_factory = AcceptableProtocolsContextFactory(context_factory, acceptable_protocols)
+        self._endpoint_factory = _StandardEndpointFactory(
+            self._reactor, self._context_factory,
+            connect_timeout, bind_address
+        )
+
+    def get_endpoint(self, uri: URI):
+        return self._endpoint_factory.endpointForURI(uri)
+
+    def negotiate(self, uri: URI) -> Deferred:
+        """Returns a deferred which fires with the transport """
+        d = Deferred()
+        endpoint = self.get_endpoint(uri)
+        factory = NegotiateProtocolFactory(self._acceptable_protocols, d)
+        endpoint.connect(factory)
+        return d
+
+
+class HTTPNegotiateDownloadHandler:
+    def __init__(self, settings: Settings, crawler=None):
+        self._settings = settings
+        self._crawler = crawler
+
+        self.acceptable_protocols = SUPPORTED_PROTOCOLS
+
+        # ToDo: Maybe load download-handlers from the settings?
+        #  Need to add default for 'h2' in DOWNLOAD_HANDLERS_BASE
+        self._http1_download_handler = HTTP11DownloadHandler(settings, crawler)
+        self._h2_download_handler = H2DownloadHandler(settings, crawler)
```

Comment:
Should I load download handler class from the settings?

Reply:
I don’t consider it a blocker, but yes, that would be ideal.

I imagine we want a new setting that is a dictionary where keys are protocol version strings (e.g. `http1.1`, `h2`) and values are download handler classes (or their import paths). I’m not sure how to call it, though.

Reply:
> I imagine we want a new setting that is a dictionary where keys are protocol version strings 

Should I add the setting as `HTTP_DOWNLOAD_HANDLERS_BASE`?

Reply:
Works for me. We can always rename later if someone can think of a better name.

Reply:
Okay